solana/programs/token-bridge-relayer/src/processor/initialize.rs

```diff
@@ -41,23 +58,91 @@ pub struct Initialize<'info> {
 
+    #[account(address = bpf_loader_upgradeable::ID)]
+    pub bpf_loader_upgradeable: UncheckedAccount<'info>,
+
     pub system_program: Program<'info, System>,
 }
 
-pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
-    let owner = ctx.accounts.owner.key();
+pub fn initialize<'a, 'b, 'c, 'info>(
+    ctx: Context<'a, 'b, 'c, 'info, Initialize<'info>>,
+    fee_recipient: Pubkey,
+    admins: Vec<Pubkey>,
+) -> Result<()> {
+    msg!(
+        "Authority: {:?}",
+        ctx.accounts.program_data.upgrade_authority_address
+    );
+    //We only update the upgrade authority if the program wasn't deployed by the designated owner
+    if Some(ctx.accounts.owner.key()) != ctx.accounts.program_data.upgrade_authority_address {
+        msg!("Setting program authority to the owner...");
+        //This call fails for anyone but the deployer who must be the current update authority.
+        invoke(
+            &bpf_loader_upgradeable::set_upgrade_authority(
+                &ctx.program_id,
+                &ctx.accounts.deployer.key(),
+                Some(&ctx.accounts.owner.key()),
+            ),
+            &[
+                ctx.accounts.program_data.to_account_info(),
+                ctx.accounts.deployer.to_account_info(),
+                ctx.accounts.owner.to_account_info(),
+            ],
+        )?;
+        msg!("Program authority set to the owner");
+    }
 
-    *ctx.accounts.tbr_config = TbrConfigState {
-        owner,
+    ctx.accounts.tbr_config.set_inner(TbrConfigState {
+        owner: ctx.accounts.owner.key(),
         pending_owner: None,
-        fee_recipient: owner,
+        fee_recipient,
         evm_transaction_size: 0,
         evm_transaction_gas: 0,
         sender_bump: ctx.bumps.wormhole_sender,
         redeemer_bump: ctx.bumps.wormhole_redeemer,
         bump: ctx.bumps.tbr_config,
-    };
-    ctx.accounts.admin_badge.bump = ctx.bumps.admin_badge;
+    });
+
+    ctx.accounts.auth_badge.set_inner(AuthBadgeState {
+        address: ctx.accounts.owner.key(),
+    });
+
+    require_eq!(
+        admins.len(),
+        ctx.remaining_accounts.len(),
+        TokenBridgeRelayerError::AdminCountMismatch
+    );
+
+    for (admin, badge_acc_info) in zip(admins, ctx.remaining_accounts) {
+        let bump = Pubkey::find_program_address(
+            &[AuthBadgeState::SEED_PREFIX, admin.to_bytes().as_ref()],
+            ctx.program_id,
+        )
+        .1;
+
+        anchor_lang::system_program::create_account(
+            CpiContext::new_with_signer(
+                ctx.accounts.system_program.to_account_info(),
+                anchor_lang::system_program::CreateAccount {
+                    from: ctx.accounts.deployer.to_account_info(),
+                    to: badge_acc_info.clone(),
+                },
+                &[&[
+                    AuthBadgeState::SEED_PREFIX,
+                    admin.to_bytes().as_ref(),
+                    &[bump],
+                ]],
+            ),
```

Comment:
This could have used [create_account_with_seed](https://docs.rs/solana-program/latest/solana_program/system_instruction/fn.create_account_with_seed.html) ([solana code](https://github.com/solana-labs/solana/blob/27eff8408b7223bb3c4ab70523f8a8dca3ca6645/programs/system/src/system_processor.rs#L336)), no? (I failed to realize this in my hasty impl too.)